tutorials/basic-topics/sorting/merge-sort.md

```diff
@@ -11,4 +11,521 @@ keywords:
+        nums[k] = rightArray[j] 
+        j+=1
+        k+=1
+
+```
+
+</TabItem>
+<TabItem value="java" label="Java">
+
+```java
+// call mergeSort(nums , 0 ,nums.length -1) in public
+
+private void mergeSort(int[] nums, int l, int r) {   
+    // nums -> ArrayList , l -> starting index , r-> end of the ArrayList
+    if (l >= r) return;
+    int mid = l + (r - l) / 2;
+    mergeSort(nums, l, mid);
+    mergeSort(nums, mid + 1, r);
+    merge(nums, l, r);
+}
+
+private void merge(int[] nums, int l, int r) {
+    int mid = l + (r - l) / 2;
+    int[] tmp = new int[r - l + 1];
+    int i = l, j = mid + 1, k = 0;
+    while (i <= mid || j <= r) {
+        if (i > mid || j <= r && nums[i] > nums[j]) {
+            tmp[k++] = nums[j++];
+        } else {
+            tmp[k++] = nums[i++];
+        }
+    }
+}
+```
+</TabItem>
+</Tabs>
+
+## Bottom Up Approach / Iterative technique
+
+### Algorithm 
+
+```
+1. So it  starts with an  element in the array 
+
+    It is an iterative approach and because one item array is always sorted
+ 
+2. Compares two nearby elements to merge into a sorted sub array 
+
+    Similarly, we then merge the sorted subarrays like we have done in top-down recursive approach (two-pointer approach) 
+
+3. Continues until we have a sorted array 
+```
+<Tabs>
+<TabItem value="java" label="Java">
+
+```java
+class Solution {
+    public List<Integer> sortArray(int[] nums) {
+        List<Integer> res = new ArrayList<>();
+        if (nums == null || nums.length == 0) return res;
+        mergeSort2(nums);
+        for (int i : nums) res.add(i);
+        return res;
+    }
+
+    private void mergeSort2(int[] nums) { // iterative only 
+        for (int size = 1; size < nums.length; size *= 2) {
+            for (int i = 0; i < nums.length - size; i += 2 * size) {
+                int mid = i + size - 1;
+                int end = Math.min(i + 2 * size - 1, nums.length - 1);
+                merge2(nums, i, mid, end);
+            }
+        }
+    }
+
+    private void merge2(int[] nums, int l, int mid, int r) {
+        int[] tmp = new int[r - l + 1];
+        int i = l, j = mid + 1, k = 0;
+        while (i <= mid || j <= r) {
+            if (i > mid || j <= r && nums[i] > nums[j]) {
+                tmp[k++] = nums[j++];
+            } else {
+                tmp[k++] = nums[i++];
+            }
+        }
+        System.arraycopy(tmp, 0, nums, l, r - l + 1);
+    }
+}
+```
+</TabItem>
+</Tabs>
+
+**Time Complexity** : Best & Worst & Average is   $O(n \log_{2} n)$
+
+**Space Complexity** :  $O(n)$ 
+
+**Stablitiy**       :  Stable
+
+**Is In Place**      :  Not-In-Place
+
+**When  not to use**: When the array is too large
+
+
+
+  
+Merging of n elements takes n time and since each time the array is cut into half it takes   $\log_{2} n$   time to reach the top so **total time complexity is**    $O(n\log_{2} n)$.
+
+
+![image](https://cdn.kastatic.org/ka-perseus-images/5fcbebf66560d8fc490de2a0d8a0e5b1d65c5c54.png)
```

Comment:
if you don't make this image, then don't use it. or give credit

Reply:
I guess the guy who has used this image also picked it up from somewhere! 
Still, I will give the credits but the he doesn't deserve it I guess